src/operations/operation.ts

```diff
@@ -61,6 +62,12 @@ export abstract class AbstractOperation<TResult = any> {
 
   options: OperationOptions;
 
+  /** @internal */
+  timeout?: Timeout;
```

Comment:
I'm confused about the purpose of the `timeout`.  I commented in `Server.ts`, but at no point that I know of will we ever use a generic "CSOT" timeout.  Instead, at each point where we apply CSOT's logic to async operations, we will create custom timeouts that apply CSOT logic (i.e., a timeout that applies to socket read specifically or socket write).  Am I misunderstanding?

Given that, it feels odd to create a promise that we won't use.  And it seems odd to use our generic Timeout class to store CSOT-specific context (`Timeout.min`, `Timeout.remainingTime`).

Thoughts on a generic `CSOTContext` class?  That's basically what the `TimeoutControllerFactory` was in the initial design.  Something that holds CSOT context, is instantiated at operation construction time, and generates timeouts for certain parts of the driver (i.e., `context.timeoutForServerSelection()`, which encapsulates CSOT logic for server selection into a single, unit testable class).

(this relates to my other comment about consolidating timeout logic for server selection - we'll need it eventually for change streams too so it makes sense not to implement it three times).

Reply:
The `CSOTContext` class sounds somewhat similar to the `OperationContext` idea that you had explored earlier on in the CSOT design. I'm not against it on principle, but I think we scrapped that idea at some point. Will look for any documents we had on why and report back.

Reply:
Just to close the loop here: We do use the "csot" timeout when it is the lower bound of time remaining before the operation is meant to expire. Server selection and connection checkout need to never take more than `serverSelectionTimeoutMS` but can be given less time if a csot timeout is smaller than that setting. So the logic in this PR does some gte/lte checks and determines if we need to make a new timer to maintain that requirement.

Reply:
> The CSOTContext class sounds somewhat similar to the OperationContext idea that you had explored earlier on in the CSOT design. I'm not against it on principle, but I think we scrapped that idea at some point. Will look for any documents we had on why and report back.

The context would be similar to the timeout factory, not the operation context.  We decided not to implement an operation context because we can pass CSOT-related data on the options objects in the driver.  I'm proposing a CSOT context to encapsulate CSOT logic, which we would then pass through the driver on the options objects.

---
I do not consider this thread resolved.  I think the current implementation is more complicated than necessary because we only sometimes re-use the timeout.

- This means server selection and connection checkout must be responsible for determining when to reuse the timeout.
- We can't unconditionally clear the timeout because it may be re-used later.

Instead, I propose we either:

1. Always create a new timeout for server selection and connection checkout.
3. Always reuse the same Timeout object, but just reset the Timeout's interval for server selection and connection checkout.

Regardless of which is chosen, I think the resultant code is simpler because server selection and connection checkout 1) do not worry about whether or not they need to use a cached timeout or create a new one 2) they can always clear the timeout.

This works especially nicely with the TimeoutFactory or a TimeoutContext, because we can encapsulate all timeout related logic into a single place that's easily unit testable.  I'm partial to the factory approach:

```typescript
class TimeoutFactory {
  private timeoutMS: number | null;
  private started = now();

  getTimeoutForServerSelection(): Timeout {
    // returns a timeout, handling CSOT vs Legacy timeout logic
  }
}

class Topology {

  selectServer(options: { ..., timeoutFactory: TimeoutFactory }) {
    ...
    const timeout = timeoutFactory.getTimeoutForServerSelection();
    try {
		....
	} finally {
      timeout.clear();
    }
  }
}
```

Note that with an approach like this, whether or not we reuse a timeout can easily be encapsulated into the TimeoutFactory by instantiating a timeout when the factory is constructed and returning the cached timeout where needed.

But a context class could suffice too:

```typescript
class TimeoutContext {
  private timeoutMS: number | null;
  private started = now();

  getTimeoutForServerSelection(): number {}
}


class Topology {

  selectServer(options: { ..., timeoutContext: TimeoutContext }) {
    ...
    const timeout = Timeout.expires(timeoutFactory.getTimeoutForServerSelection());
    try {
		....
	} finally {
      timeout.clear();
    }
  }
}
```

An approach like this consolidates CSOT logic and can be reused outside of the main code path (i.e., topology connect).  

I don't think this work needs to block this PR.  But I do want to make sure we discuss this, and I'd like to consider one of these approaches in a future ticket.